text/0000-rvalue_static_promotion.md

```diff
@@ -0,0 +1,189 @@
+- Feature Name: rvalue_static_promotion
+- Start Date: 2015-12-18
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Promote constexpr rvalues to values in static memory instead of
+stack slots, and expose those in the language by being able to directly create
+`'static` references to them. This would allow code like
+`let x: &'static u32 = &42` to work.
+
+# Motivation
+[motivation]: #motivation
+
+Right now, when dealing with constant values, you have to explicitly define
+`const` or `static` items to create references with `'static` lifetime,
+which can be unnecessarily verbose if those items never get exposed
+in the actual API:
+
+```rust
+fn return_x_or_a_default(x: Option<&u32>) -> &u32 {
+    if let Some(x) = x {
+        x
+    } else {
+        static DEFAULT_X: u32 = 42;
+        &DEFAULT_X
+    }
+}
+fn return_binop() -> &'static Fn(u32, u32) -> u32 {
+    const STATIC_TRAIT_OBJECT: &'static Fn(u32, u32) -> u32
+        = &|x, y| x + y;
+    STATIC_TRAIT_OBJECT
+}
+```
+
+This workaround also has the limitation of not being able to refer to
+type parameters of a containing generic functions, eg you can't do this:
+
+```rust
+fn generic<T>() -> &'static Option<T> {
+    const X: &'static Option<T> = &None::<T>;
+    X
+}
+```
+
+Additionally, despite it being memory safe, it is not currently possible to
+create a `&'static mut` to a zero-sized type without involving unsafe code:
+
+```rust
+fn return_fn_mut_or_default(&mut self) -> &FnMut(u32, u32) -> u32 {
```

Comment:
Is this signature a typo? In particular, did you mean to write that the return value is a `&mut FnMut(u32, u32) -> u32` ?